_posts/2019-05-28-improved-hive-bucketing.md

```diff
@@ -0,0 +1,135 @@
+---
+layout: post
+title:  "Improved Hive Bucketing"
+author: David Phillips
+---
+
+The next release of Presto (version 312) adds support for the more
+flexible bucketing introduced in recent versions of Hive. Specifically,
+it allows any number of files per bucket, including zero. This allows
+inserting data into an existing partition without having to rewrite the
+entire partition, and improves the performance of writes by not requiring
+the creation of files for empty buckets.
+
+# Hive bucketing overview
+
+Hive bucketing is a simple form of hash partitioning. A table is bucketed
+on one or more columns with a fixed number of hash buckets:
+
+```sql
+CREATE TABLE page_views (
+  user_id bigint,
+  page_url varchar,
+  dt date
+)
+WITH (
+  partitioned_by = ARRAY['dt'],
+  bucketed_by = ARRAY['user_id'],
+  bucket_count = 50
+)
+```
+
+The bucketing happens within each partition of the table. In the above
+example, the table is partitioned by date and is declared to have `50` buckets
+using the user ID column. This means that the table will have `50` buckets
+for each date. The assigned bucket for each row is determined by hashing
+the user ID value.
+
+# Original Hive bucketing
+
+Originally, Hive required exactly one file per bucket. The files were named 
```

Comment:
Do you know the version for the "originally"?

Reply:
Unfortunate not, it's hard to track exactly what changed when. This optimization seems to have been introduced for Tez, but I'm not confident enough to say here.

Reply:
Yup, fingers point at Tez.